src/Features/CSharp/Portable/BraceCompletion/AbstractCurlyBraceOrBracketCompletionService.cs

```diff
@@ -0,0 +1,273 @@
+
+                using var _ = ArrayBuilder<TextChange>.GetInstance(out var mergedChanges);
+                var amountToShift = 0;
+                foreach (var newRange in newRanges)
+                {
+                    var newTextChangeSpan = newRange.Span;
+                    // Get the text to put in the text change by looking at the span in the formatted text.
+                    // As the new range start is relative to the original text, we need to adjust it assuming the previous changes were applied
+                    // to get the correct start location in the formatted text.
+                    // E.g. with changes
+                    //     1. Insert "hello" at 2
+                    //     2. Insert "goodbye" at 3
+                    // "goodbye" is after "hello" at location 3 + 5 (length of "hello") in the new text.
+                    var newTextChangeText = formattedText.GetSubText(new TextSpan(newRange.Span.Start + amountToShift, newRange.NewLength)).ToString();
+                    amountToShift += (newRange.NewLength - newRange.Span.Length);
+                    mergedChanges.Add(new TextChange(newTextChangeSpan, newTextChangeText));
+                }
+
+                return mergedChanges.ToImmutable();
+            }
+        }
+
+        /// <summary>
+        /// Formats the span between the opening and closing points, options permitting.
+        /// Returns the text changes that should be applied to the input document to 
+        /// get the formatted text and the end of the close curly brace in the formatted text.
+        /// </summary>
+        private async Task<(ImmutableArray<TextChange> textChanges, int finalBraceEnd)> FormatTrackingSpanAsync(
+            Document document,
+            int openingPoint,
+            int closingPoint,
+            bool shouldHonorAutoFormattingOnCloseBraceOption,
+            ImmutableArray<AbstractFormattingRule> braceFormattingIndentationRules,
+            DocumentOptionSet documentOptions,
+            CancellationToken cancellationToken)
+        {
+            var option = document.Project.Solution.Options.GetOption(BraceCompletionOptions.AutoFormattingOnCloseBrace, document.Project.Language);
+            if (!option && shouldHonorAutoFormattingOnCloseBraceOption)
+            {
+                return (ImmutableArray<TextChange>.Empty, closingPoint);
+            }
+
+            // Annotate the original closing brace so we can find it after formatting.
+            document = await GetDocumentWithAnnotatedClosingBraceAsync(document, closingPoint, cancellationToken).ConfigureAwait(false);
+
+            var text = await document.GetTextAsync(cancellationToken).ConfigureAwait(false);
+            var root = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+
+            var startPoint = openingPoint;
+            var endPoint = AdjustFormattingEndPoint(text, root, startPoint, closingPoint);
```

Comment:
The code in `CurlyBraceCompletionService.AdjustFormattingEndPoint` was previously inline at this point